LOGIN_FLOW.md

```diff
@@ -0,0 +1,355 @@
+    service worker that has access to the contract state (so that it can decide
+    when a notificaton is relevant).
+
+### Constraints of service workers
+
+To understand some of the decisions made, it's important to understand how
+web (service) workers function and some of their constraints.
+
+Web workers run in an entirely different context from the browsing context, and
+by default they don't have access to the browsing context (and vice versa). This
+means that, for example, `self.foo = 'bar'` executed in a browsing context will
+not make the global variable `foo` available in a web worker, and, conversely,
+the same code executing in a web worker will not make it available in the
+browsing context.
+
+Instead, communication between web workers and browsing contexts is asynchronous
+and event-driven. The most relevant web APIs for how this communication happens
+are `postMessage`, the `message` event listeners (e.g., `onmessage`) and the
+`structuredClone` function.
+
+Following up the previous example, if a web worker wanted to communicate to the
+browsing context that `foo` should take the value of `bar`, the process is more
+involved than just `self.foo = 'bar'`. Instead, the process looks more or less
+like this:
+
+  1. The browsing context sets up an event listener for the `message` event on
+     the worker.
+  2. The worker sends a message event with the new value. Under the hood,
+     message data will be processed using the structure clone algorithm (i.e.,
+     `structuredClone`).
+  3. The event listener receives and processes the event.
+
+#### Example
+
+```js
+console.log('foo is', self.foo)
+// > foo is undefined
+const worker = new Worker('data:text/javascript,' + encodeURIComponent(`
+  // The worker will send a message to the browsing context
+  postMessage(['foo', 'bar'])
+`))
+worker.onmessage = function (event) {
+  const [key, value] = event.data
+  self[key] = value
+}
+setTimeout(() => {
+  console.log('foo is', self.foo)
+  // > foo is bar
+}, 100)
+```
+
+#### The structured clone algorithm
+
+Explaining how the structured clone algorithm works in detail is complex.
+However, the short overview is that a _very close approximatiion_ is the
+equivalent of JSON. I.e., it's something like this:
+
+```js
+function structuredClone (data) {
+  return JSON.parse(JSON.stringify(data))
+}
+```
+
+Although the structured data clone algorithm supports more types than plain
+JSON, the data types supported are predefined, and, importantly, functions and
+custom object types are _not_ supported. The `serdes` library in
+`shared/domains/chelonia` is a collection of functions to augment the
+capabilities of the structured clone algorithm.
+
+#### Specifics about Service Workers
+
+Unlike other types of web workers, there can be a instance of single service
+worker for any given page. A single service worker can be shared across multiple
+browsing contexts (e.g., if the user opens many tabs with Group Income, they
```

Comment:
Indeed. That sentence needs correcting because it looks like some editing went wrong. However,

> worker per website

That's inaccurate, depending on how you define website. You can also have multiple service workers defined (e.g., during upgrades), but at most a single one can be active.

Reply:
This is how they say a similar thing on MDN:

> A single service worker can control many pages. Each time a page within your scope is loaded, the service worker is installed against that page and operates on it.